semantic-core/src/Analysis/Eval.hs

```diff
@@ -41,12 +42,12 @@ eval Analysis{..} eval = \case
       addr <- alloc n
       v <- bind n addr (eval (instantiate1 (pure n) b))
       v <$ assign addr v
-    a :>> b -> eval a >> eval b
+    a :>> b -> (<>) <$> eval a <*> eval b
```

Comment:
Combining the results of evaluations together with `<>` enables us to model effects compositionally. E.g. given a program like:

```ruby
def f
  a
  b
end
```

Then a domain modelling the type-and-effect of `f` will define `<>` to combine the effects incurred by both `a` and `b` while taking the type of only `b`, whereas a domain modelling the import graph would union together any imports performed by either `a` or `b`. On the other hand, the concrete domain implements `<>` to simply discard `a`, since it isn’t attempting to instrument the effects, only perform them.

Reply:
It’s also worth asking: why a semigroup, when so much of the abstract interpretation literature focuses on (semi)lattices? In short, our lattice structure is provided by the sets of computations performed under the nondeterminism effect; paraphrasing Darais et al, concretization maps values in the abstract domain to _finite sets of observations_ in the concrete domain. So in short, the lattice’s least upper bound operation is a) provided by `NonDet`, and b) semantically distinct from the use of `<>` to combine sequenced abstracted values.

Reply:
Finally, we could eventually consider strengthening the `Semigroup` constraint to `Monoid` to allow us to represent divergent/exceptional computation with `mempty`; or further strengthening it to some sort of unital semiring structure to combine the two kinds of combination, with addition for the least upper bound, zero for the bottom, multiplication for sequencing, and `unit` for its, well, unit. (I haven’t actually checked to see if this would actually be a lawful semiring; the distributive laws seem like they’d be pretty interesting.)

Reply:
This all makes total sense to me, and is a huge win for clarity. Can we move the relevant sections of the above comments (which are wonderful!) to comments on the `deriving` instances for `Concrete` and `Term Monotype a`?